yt/frontends/ramses/io_utils.pyx

```diff
@@ -0,0 +1,151 @@
+def read_amr(FortranFile f, dict headers,
+             np.ndarray[np.int64_t, ndim=1] ngridbound, int min_level,
+             RAMSESOctreeContainer oct_handler):
+
+    cdef int ncpu, nboundary, max_level, nlevelmax, ncpu_and_bound
+    cdef double nx, ny, nz
+    cdef int ilevel, icpu, ng, n
+    cdef np.ndarray[np.int32_t, ndim=2] numbl
+    cdef np.ndarray[np.float64_t, ndim=2] pos
+
+    numbl = headers['numbl']
+    nboundary = headers['nboundary']
+    nx, ny, nz = (((i-1.0)/2.0) for i in headers['nx'])
+    nlevelmax = headers['nlevelmax']
+    ncpu = headers['ncpu']
+
+    ncpu_and_bound = nboundary + ncpu
+
+    pos = np.empty((0, 3), dtype=np.float64)
+    for ilevel in range(nlevelmax):
+        for icpu in range(ncpu_and_bound):
+            if icpu < ncpu:
+                ng = numbl[ilevel, icpu]
+            else:
+                ng = ngridbound[icpu - ncpu + nboundary*ilevel]
+
+            if ng == 0:
+                continue
+            # ind = fpu.read_vector(f, "I").astype("int64")  # NOQA
+            f.skip(3)
+
+            # Reallocate memory if requred
+            if ng > pos.shape[0]:
+                pos = np.empty((ng, 3), dtype="d")
+
+            pos[:ng, 0] = f.read_vector("d") - nx
+            pos[:ng, 1] = f.read_vector("d") - ny
+            pos[:ng, 2] = f.read_vector("d") - nz
+
+            f.skip(31)
+            # Note that we're adding *grids*, not individual cells.
+            if ilevel >= min_level:
+                n = oct_handler.add(icpu + 1, ilevel - min_level, pos[:ng, :],
+                                    count_boundary = 1)
+                if n > 0:
+                    max_level = max(ilevel - min_level, max_level)
+
+    return max_level
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+@cython.nonecheck(False)
+cpdef read_offset(FortranFile f, int min_level, int domain_id, int nvar, dict headers):
+
+    cdef np.ndarray[np.int64_t, ndim=1] offset, level_count
+    cdef int ndim, twotondim, nlevelmax, n_levels, nboundary, ncpu, ncpu_and_bound
+    cdef int file_ilevel, file_ncache, ilevel, icpu, skip_len
+
+    numbl = headers['numbl']
+    ndim = headers['ndim']
+    nboundary = headers['nboundary']
+    nlevelmax = headers['nlevelmax']
+    n_levels = nlevelmax - min_level
+    ncpu = headers['ncpu']
+
+    ncpu_and_bound = nboundary + ncpu
+    twotondim = 2**ndim
+
+    skip_len = twotondim * nvar
+
+    # It goes: level, CPU, 8-variable (1 cube)
+    offset = np.zeros(n_levels, dtype=np.int64)
+    offset -= 1
+    level_count = np.zeros(n_levels, dtype=np.int64)
+
+    cdef long[:] level_count_view = level_count
+    cdef long[:] offset_view = offset
+
+    for ilevel in range(nlevelmax):
+        for icpu in range(ncpu_and_bound):
+            file_ilevel = f.read_int()
+            file_ncache = f.read_int()
+            if file_ncache == 0: continue
+
+            if file_ilevel != ilevel+1:
+                raise Exception()
```

Comment:
Can this be raised with an error message? Also unless you're immediately catching it further up the stack, raising the base `Exception` class is usually discouraged in favor of using a subclass that conveys what the problem is.